### PR TITLE
Introduce the ResultIterator

### PR DESCRIPTION
**Pull request by withrocks**
The `ResultIterator` behaves much like the part of a Django `QuerySet` that
deals with iteration. We use it to iterate over results of thick models,
which are based on Django models.

### CODE REVIEW

File: src/clims/services/base_extensible_service.py [hunk 1]
```diff
@@ -61,24 +62,24 @@ def all(self):
         # which in turn would return the wrapper when materialized. For that to work smoothly,
         # we'll need to look into implementation details of django querysets.
 
-        # TODO: how does the prefetch perform when fetching all objects like this
-        for entry in self._all_qs():
-            yield self.to_wrapper(entry)
+        # TODO: The `to_wrapper` method is NOT on this mixin, which makes this difficult to reason
```
Comment (johandahlberg) [inline src/clims/services/base_extensible_service.py:65]: I agree with this. Do you think we should fix it in this PR, or do it later?

**Review by johandahlberg — APPROVED** [review]
This looks very nice and clean to me. Good job! Do you want to fix the TODO I commented on now? Or wait and do it later?

File: src/clims/services/base_extensible_service.py (hunk 1 above)
Comment (withrocks) [reply src/clims/services/base_extensible_service.py:65]: I think I would like to leave it for later, as I has a ton of PRs to get in :|

**Comment by withrocks** [comment]
Cool, thanks! Please merge it.

**Pull request merged**